Add OPT use case

palec87 opened the pull request:
**Main contributions:**

- OPT controller: imswitch\imcontrol\controller\controllers\ScanControllerOpt.py
- OPT widget: imswitch\imcontrol\view\widgets\ScanWidgetOpt.py
- Alignment controller: imswitch\imcontrol\controller\controllers\AlignOptController.py
- Alignment widget: imswitch\imcontrol\view\widgets\AlignOptWidget.py

**Other notables**
- added ic4 TIS camera support, `imagingcontrol4`
- added arduino support with `telemetrix` library
- proposed `ScanExecutionMonitor` class in `ScanControllerOpt.py`

Tested with NEMA17 4phase stepper, DMK252 TIS camera
OS: Win

Other tests under way with Hamamatsu ORca.

palec87 committed:
init opt experiment with arduino stepper

palec87 committed:
Corrections and Align COR widget

jacopoabramo committed:
Added CommChannel signals to ScanControllerOpt

jacopoabramo committed:
Added further comments on implementation

jacopoabramo committed:
Added live view and memory recording signals

jacopoabramo committed:
Added live view signals

jacopoabramo committed:
Connected new signals to CommChannel

jacopoabramo committed:
Added stub connections to CommChannel

jacopoabramo committed:
Added missing flag

palec87 committed:
Merge pull request #1 from jacopoabramo/comm-channel-patch

CommunicationChannel integration for dynamic OPT scan control

palec87 committed:
stashed changes

palec87 committed:
opt controller, fixed signals and snapping

palec87 committed:
disconnect MemorySnapAvailable before OPT, only workaround

palec87 committed:
add rotator selector to OPT scanController, widget reorganized

palec87 committed:
OPT step divisibility check

palec87 committed:
cannot fix mockstepper not updating image stack

jacopoabramo committed:
Control over snap visualization

palec87 committed:
align opt and camera depth unsuccessful fixes

palec87 committed:
ic4 tis manager, works minus the switching between pixel depths.

palec87 committed:
switching depth by reconnecting camera does not work

palec87 committed:
working ic4 v0

palec87 committed:
remove prints, fixed corrections, better alignOpt

palec87 committed:
debugging, cleaning widget features, removing debug codes.

palec87 committed:
improved snapping, still crashing sometimes on pixel_format change.

palec87 committed:
cleane refactored, local acquisition loop, some docs

palec87 committed:
delete logs and prints, change save fname for corrections

palec87 committed:
clean, camera buffer fix, convert widgets to spinboxes

palec87 committed:
bug fixing, step_opt_done not resolved

palec87 committed:
correction

jacopoabramo committed:
Adding sigOptStepDone for rotators

palec87 committed:
moving arduino stepper, mock not working yet

palec87 committed:
fix sigOpt bug

palec87 committed:
fix mock rotator, update stack still not working however

palec87 committed:
if/else to current position callback, but does not fix the update

palec87 committed:
mock viewer still not updating

palec87 committed:
time reporting class

jacopoabramo committed:
Refactoring `ArduinoRotatorManager`

- cleaning up dead code;
- removing signal connections;
- started base for checks on manager config data;
- removed old interface;
- added implementation for `MockBoard`
- added dependency of `APSCheduler`

jacopoabramo committed:
Added base examples for OPT setup

jacopoabramo committed:
Moved MockBoard to telemetrix.py

jacopoabramo committed:
Extending `__init__`

jacopoabramo committed:
Typo

jacopoabramo committed:
Typos

jacopoabramo committed:
Setting stepsPerTurn before motor init

jacopoabramo committed:
Add docstring

jacopoabramo committed:
Add signal to notify of new current position

jacopoabramo committed:
Typo fix

jacopoabramo committed:
Typo fix

jacopoabramo committed:
Changes to adapt for correct scanning

jacopoabramo committed:
Refactoring

jacopoabramo committed:
Renaming `MockBoard` to `MockTelemetrixBoard`

jacopoabramo committed:
Further refactoring

jacopoabramo committed:
Signal renaming

jacopoabramo committed:
Continuing refactor

jacopoabramo committed:
Move yes/no question to view

jacopoabramo committed:
Moving widget initialization to `__post_init__`

jacopoabramo committed:
Continuing refactoring

jacopoabramo committed:
Continue refactoring

jacopoabramo committed:
Continuing refactor

jacopoabramo committed:
File renaming

jacopoabramo committed:
File renaming

jacopoabramo committed:
Fixing data tuple index

- Must be reviewed how data is received from telemetrix interface

jacopoabramo committed:
Fixing callback invocation

jacopoabramo committed:
Removing prints

jacopoabramo committed:
Continuing refactor

- mock experiment with synthetic data works
- report for mock data works

jacopoabramo committed:
Removing old code

jacopoabramo committed:
Adjusting some lines

jacopoabramo committed:
Unused import

palec87 committed:
further refactoring, comments, replies

jacopoabramo committed:
Some comments

jacopoabramo committed:
Update copyright

jacopoabramo committed:
Added separate method for synthetic data retrieval

jacopoabramo committed:
Small docstring change

jacopoabramo committed:
Removing comments

jacopoabramo committed:
Refactor to use `QDialog` for report plot window

jacopoabramo committed:
Manager renaming

- from `ArduinoRotatorManager` to `TelemetrixRotatorManager`;
- Telemetrix supports different boards; it makes it more
familiar for users to integrate it

jacopoabramo committed:
Specifying private method

jacopoabramo committed:
Some typo

jacopoabramo committed:
Added `TODO`

jacopoabramo committed:
Adding TODOs

palec87 committed:
fixed real experiment, not yet moct

palec87 committed:
fixes

palec87 committed:
test

palec87 committed:
fixed real experiment, mock still does not update after each step

jacopoabramo committed:
Docstring formatting

jacopoabramo committed:
Renaming mock response variable

- just for a little clarity

jacopoabramo committed:
Adding missing values to initialized tuple

jacopoabramo committed:
Renaming flag for clarity

- also some comments cleanup

jacopoabramo committed:
Add UI adjustment

jacopoabramo committed:
Typo fix

jacopoabramo committed:
Removed unused flag

jacopoabramo committed:
Fixing log print

jacopoabramo committed:
Added daemon thread

- the mock board initially ran on the UI thread;
- this impeded live-update of the image viewer;
- with the daemon thread the computation of the new position is offloaded from the UI

jacopoabramo committed:
Adjusting continous mock motor movement

- new position is not correctly updated... why?

jacopoabramo committed:
Docstring adjustment

jacopoabramo committed:
Removing unnecessary print

jacopoabramo committed:
Added comment

jacopoabramo committed:
Small docstring rework

jacopoabramo committed:
Added sinogram progress bar

palec87 committed:
fixed live reconstruction from the worker

palec87 committed:
refactor live recon, messages and debug

palec87 committed:
metadata, delay before getLastFrame

palec87 committed:
fixed shared attributes and mocking, confirmed ic4driver working

palec87 committed:
worker implemented for the alignOPT procedure.

palec87 committed:
refactoring and initial docs commit

jacopoabramo committed:
Fixing pyqtgraph version to 0.13.3

- newer version breaks software

jacopoabramo committed:
Adding check over setupInfo

- if no rotator is present connection should be skipped

jacopoabramo committed:
Removing unused code

jacopoabramo committed:
Adding `getExposure` API

palec87 committed:
remove stability timeMonitoring, because it is really fast.
2. wait constant set to None, and needs to be set in the prepareOptScan

palec87 committed:
Merge branch 'add-camera-mock' of https://github.com/palec87/ImSwitchOpt into add-camera-mock

palec87 committed:
remove unused import

palec87 committed:
detectors list for opt from the forOpt config file flag
widget enabling

palec87 committed:
shorten exposure button

palec87 committed:
forOpt setup info support

palec87 committed:
forOpt setup info support

palec87 committed:
fixing detector list update in the widget

palec87 committed:
cosmetics

palec87 committed:
ic4 folder not needed

palec87 committed:
0. remove try except to rely on getExposure method
1. shared attributes from camera in the postScanEnd collection
2. conversion work from base of us

palec87 committed:
cosmetics

palec87 committed:
enum for Exposure to us conversion

palec87 committed:
ExposureTimeToUs implemented for tis4camera

palec87 committed:
cosmetic changes to camera managers and interfaces.

palec87 committed:
updated getExposure methods to accomodate units the submanagers use

palec87 committed:
reorder methods, extended docstrings

palec87 committed:
refactor, docstrings, add tabs to widgets, add tooltips

palec87 committed:
progress on docs

palec87 committed:
somehow there are still issues with adjusting snapping timout when linked to the exposure. Maybe a dtype issue.

palec87 committed:
docs for the OPT use case.

palec87 committed:
comment-out the timeout print

palec87 committed:
rename because of pytest

palec87 committed:
opt tests initiated

palec87 committed:
new json, plus increased timeout for tis4Manager

palec87 committed:
making sure of the np array input to the FBPliveRecon

palec87 committed:
functioning test as a blueprint

palec87 committed:
rename file

palec87 committed:
implemented frame_rate control for the snap timeout, added docstrings

palec87 committed:
cosmetics

palec87 committed:
backpedal on frame_rate settings, todo: FR convert to editable.

palec87 committed:
added imagingcontrol4 library

palec87 committed:
frame_rate and update parameters fixes

palec87 committed:
docs update, fixed links

palec87 committed:
remove prints

palec87 committed:
cosmetics

jacopoabramo commented:
Build is failing due to this:

```
ERROR: Ignored the following versions that require a different python version: 0.4.2 Requires-Python >=3.9,<3.12; 0.4.3 Requires-Python >=3.9,<3.12; 0.4.4 Requires-Python >=3.9,<3.13
ERROR: Could not find a version that satisfies the requirement imagingcontrol4>=1.0.1.2418 (from imswitch) (from versions: none)
ERROR: No matching distribution found for imagingcontrol4>=1.0.1.2418
```

It might be due to the fact that this package is not available on Python version 3.8 and 3.9. @palec87 could you double check if this is the case?

palec87 committed:
removeing ic4 library

palec87 commented:
Hi @jacopoabramo, I pushed fixes, ic4 library is only available for windows, so I removed it from `setup.cfg` file.

jacopoabramo commented on imswitch/imcontrol/model/managers/rotators/TelemetrixRotatorManager.py:
```diff
@@ -0,0 +1,222 @@
+        Args:
+            angle (int): rotation angle.
+        """
+        # TODO: move this check to controller
+        if inSteps:
+            steps = value
+        else:
+            steps = angleToSteps(value, self._stepsPerTurn)
```
Can this check be moved to the controller as stated in the comment? This would provide an easier time dealing with a single measurement unit in the model APIs

palec87 replied:
Unfortunately, to do that, I would need to change RotatorController+RotationScanController. The default for imswitch is angles, but I think it is better to internally work in steps, which is what the Telemetrix and OPT does. Propagating that is perhaps separate PR?

jacopoabramo replied:
Sorry I forgot to reply to this. Then we'll keep it as it is for now, I don't want to overcomplicate things.

jacopoabramo commented on jdf:
What is this file? Can you remove it? @palec87

jacopoabramo commented on standa_motrot_stub.py:
Can you actually remove this file entirely? It was a refuse from a previous PR

palec87 replied:
It is not mine, but I will :)

jacopoabramo reviewed (changes requested):
Overall the PR seems quite clean. I added some questions.

Just a question: did you format the codebase using black? Just for my clarification, most of the changes were just formatting so not a real issue.

palec87 commented:
> Overall the PR seems quite clean. I added some questions.
> 
> Just a question: did you format the codebase using black? Just for my clarification, most of the changes were just formatting so not a real issue.

Hi, I did not use the `black`. What is that about?

palec87 commented on imswitch/imcontrol/model/managers/detectors/APDManager.py:
```diff
@@ -109,6 +109,9 @@ def stopAcquisitionLocal(self):
         except Exception:
             pass
 
+    def getExposure(self) -> int:
+        return 0
```
Yes, actually I was mostly wrong on this one. What is important for wait constants is frame rate and not exposure. But I think exposing the exposure time is a step in the right direction anyways.

palec87 committed:
morph opt scanner to controller + file removal and all init adjustments

palec87 committed:
additional fix of optscan morphing. Propagate changes to config jsons

palec87 commented on imswitch/imcontrol/model/managers/__init__.py:
```diff
@@ -7,6 +7,7 @@
 from .RecordingManager import RecordingManager, RecMode, SaveMode, SaveFormat
 from .SLMManager import SLMManager
 from .ScanManagerPointScan import ScanManagerPointScan
+from .ScanManagerOpt import ScanManagerOpt
```
done

palec87 commented on imswitch/imcontrol/model/managers/ScanManagerOpt.py:
I thought for the future, triggering should be added, and I thought it will go throught SuperScan

jacopoabramo replied:
Hmmm I think that triggering should still pass through the current OPT controller. Nevertheless dealing with the scanning in ImSwitch at this time requires a different thinking since its targeted for FPGAs most of the time. We'll cross that bridge when it comes to that.

palec87 commented on imswitch/imcontrol/model/interfaces/pyicic/IC_GrabberDLL.py:
```diff
@@ -17,8 +17,10 @@ class IC_GrabberDLL:
     if sys.maxsize > 2**32:
-        _ic_grabber_dll = windll.LoadLibrary(ctypes.util.find_library('tisgrabber_x64.dll'))
+        # _ic_grabber_dll = windll.LoadLibrary(ctypes.util.find_library('tisgrabber_x64.dll'))
+        _ic_grabber_dll = cdll.LoadLibrary("tisgrabber_x64.dll")
```
This is my stub or running IC3.5 on my machine. Sorry.

jacopoabramo replied:
I would say that it makes sense to use CDLL to keep it OS independent. Let's leave it as it is, we'll monitor if this makes some problems in the futurue.

palec87 committed:
remove hardcoded lib path

palec87 committed:
one more removal of hanging scanOpt file

palec87 committed:
scanopt stub

palec87 committed:
refactor tomopy shepp3d funcs to imcommon

palec87 commented on imswitch/imcontrol/model/interfaces/tis4_tester.py:
It is a local tester, so if you can remove it from the PR, please do. I did not know.

palec87 commented on .gitignore:
```diff
@@ -69,6 +69,7 @@ target/
 # Specific files
 imswitch/imcontrol/model/managers/lasers/KatanaLaserManager.py
 imswitch/imcontrol/model/managers/stands/LeicaDMIManager.py
+imswitch/imcontrol/model/interfaces/tis4_tester.py
```
removed

palec87 commented on imswitch/__main__.py:
```diff
@@ -17,9 +18,11 @@ def main():
+
         enabledModuleIds.append(enabledModuleIds.pop(enabledModuleIds.index('imscripting')))
 
+    # enabledModuleIds = ['imcontrol', 'imreconstruct', 'imscripting']
+    # modulePkgs are paths to __init__ files
```
Isn't this little helpful for the new developers, I find imswitch pretty undercommented, considering so many references across modules.

jacopoabramo replied:
I agree, we can leave it like this for now

palec87 committed:
delete comments, empty test, tester from gitignore, debugging code

palec87 commented on imswitch/imcontrol/controller/controllers/ImageController.py:
```diff
@@ -29,8 +29,16 @@ def __init__(self, *args, **kwargs):
+        self._commChannel.sigSetSnapVisualization.connect(self.setSnapVisualization)
         self._commChannel.sigSetExposure.connect(lambda t: self.setExposure(t))
 
+    def setSnapVisualization(self, status: bool) -> None:
+        if status:
+            self._commChannel.sigMemorySnapAvailable.connect(self.memorySnapAvailable)
+        else:
+            # self._commChannel.sigMemoryRecordingAvailable.disconnect()
+            self._commChannel.sigMemorySnapAvailable.disconnect()
+
```
I do not recognize this, is this really in the diff? I see it emitted in the recordings manager, looks like original codebase

palec87 committed:
remove self.name from alignOpt widget

palec87 committed:
scanControllerOpt remove commented line

palec87 committed:
unified updatePosition methods for rotator

palec87 commented on imswitch/imcontrol/controller/controllers/RotatorController.py:
```diff
@@ -34,14 +42,21 @@ def closeEvent(self):
+        current position.
+
+        Args:
+            name (str): Qt element.
```
resolved, changed to Rotator's name

palec87 commented on imswitch/imcontrol/controller/controllers/RotatorController.py:
```diff
@@ -56,10 +71,12 @@ def startContMov(self, name):
-        self._widget.updatePosition(name, pos)
+        pos = self._master.rotatorsManager[name].get_position()
+        # DP, this probably breaks for other rotators!
+        self._widget.updatePosition(name, pos[1])
```
Unified and fixed returns for `position()` method. I fixed it on the level of TelemetrixRotatorManager

palec87 committed:
additional comments removal

palec87 committed:
bit more formatting

jacopoabramo commented on imswitch/imcontrol/model/managers/__init__.py:
```diff
@@ -7,7 +7,7 @@
 from .SLMManager import SLMManager
 from .ScanManagerPointScan import ScanManagerPointScan
-from .ScanManagerOpt import ScanManagerOpt
+# from .ScanManagerOpt import ScanManagerOpt
```
Can you remove this comment?

jacopoabramo reviewed (approved):
Everything seems in order, thank you for your efforts. I'll do a double check and run the tests, then I'll push the changes.

jacopoabramo commented:
> > Overall the PR seems quite clean. I added some questions.
> > Just a question: did you format the codebase using black? Just for my clarification, most of the changes were just formatting so not a real issue.
> 
> Hi, I did not use the `black`. What is that about?

[Black ](https://github.com/psf/black) is a code formatter tool. It's pretty widespread. I don't like it too much but at this time it does not affect the PR quality. We're already in the process of adjusting code linting to a more solid approach but it'll take some time so no worries.

jacopoabramo committed:
Removed tester file

jacopoabramo committed:
Unused imports

jacopoabramo committed:
Fix import

palec87 committed:
fixing tests and wrong linting in the OptController (comma in the endexing)

palec87 committed:
Merge branch 'add-camera-mock' of https://github.com/palec87/ImSwitchOpt into add-camera-mock

palec87 committed:
cdll fix acording to agreed way

palec87 committed:
remove comment

palec87 committed:
revert of tismanager and tiscamera

jacopoabramo committed:
Updating test workflow

jacopoabramo committed:
Merge branch 'add-camera-mock' of https://github.com/palec87/ImSwitchOpt into add-camera-mock

jacopoabramo committed:
Fixing python version typo

palec87 committed:
new AlignCOR class, no widget changes yet

palec87 committed:
Merge branch 'add-camera-mock' of https://github.com/palec87/ImSwitchOpt into add-camera-mock

palec87 committed:
changed rotator separator in the widget, hard to emulate

palec87 committed:
new alignment with cumsums, does not work, but plotting canvas is there

palec87 committed:
threshold implemented

palec87 committed:
remove raw cumsums, normalize crosscorrelation to shift 0, so it can be maximized

palec87 committed:
fixed enabling elements, normalization of the cumsums

palec87 committed:
cosmetics

palec87 committed:
add modality combobox

palec87 committed:
working code, but unsure if it really captures the correction

palec87 committed:
merge image put to tabs to make more space

palec87 committed:
invert hor. cuts for the case of transmission modality

palec87 committed:
fixing plots, rounding and inversion of the horizontal cuts

palec87 committed:
refine docstrings and comments

palec87 committed:
details

palec87 committed:
works, no need for cumsums so commented out for now.

jacopoabramo committed:
typo

jacopoabramo committed:
typo

jacopoabramo committed:
Update docstring

jacopoabramo committed:
Docstring correction

palec87 committed:
refactor tomopy code, add licence, permalink and specified line codes

palec87 committed:
fixed saving bug with tiff extension

palec87 committed:
updated docs and spellcheck

palec87 committed:
update docs

palec87 committed:
removing summulative sums codes, + refactor docs

jacopoabramo committed:
Attempting to fix UTs

- some features are deprecated in latest napari versions;
- needs to be addressed in main repo

jacopoabramo commented:
I had to pin-point napari version to below 0.5.0 since there are lots of breaking changes which are incompatible with ImSwitch. These will be taken care of in the main repo.

palec87 committed:
new screenshots

palec87 committed:
Merge branch 'add-camera-mock' of https://github.com/palec87/ImSwitchOpt into add-camera-mock

jacopoabramo commented:
@palec87 I did a quick review of the latest changes. It seems you only updated the documentation so far is that correct?

palec87 commented:
> @palec87 I did a quick review of the latest changes. It seems you only updated the documentation so far is that correct?

Hi @jacopoabramo, yes last fixes regard only docs and screenshots. If you reviewed the align controller and widgets before, then everything should be alrighty

jacopoabramo commented:
@palec87 PR approved and closed. Sorry it took so long, there were a huge number of conflicts with recent merges.
